oauth2_http/java/com/google/auth/oauth2/LoggingUtils.java

```diff
@@ -0,0 +1,155 @@
+        logger.error(message);
+        break;
+      default:
+        logger.info(message);
+        // Default to INFO level
+    }
+    MDC.clear();
+  }
+
+  static void logRequest(HttpRequest request, Logger logger, String message) {
+    if (logger.isInfoEnabled()) {
+      try {
+        Map<String, String> loggingDataMap = new HashMap<>();
+        loggingDataMap.put("request.method", request.getRequestMethod());
+        loggingDataMap.put("request.url", request.getUrl().toString());
+
+        Map<String, Object> headers = new HashMap<>();
+        request
+            .getHeaders()
+            .forEach(
+                (key, val) -> {
+                  if ("authorization".equals(key)) {
+                    String hashedVal = calculateSHA256Hash(String.valueOf(val));
+                    headers.put(key, hashedVal);
+                  } else {
+                    headers.put(key, val);
+                  }
+                });
+        loggingDataMap.put("request.headers", headers.toString());
+
+        if (request.getContent() != null && logger.isDebugEnabled()) {
+          // are payload always GenericData? If so, can parse and store in json
+          GenericData data = (GenericData) ((UrlEncodedContent) request.getContent()).getData();
+
+          Map<String, String> contextMap = parseGenericData(data);
+          loggingDataMap.put("request.payload", contextMap.toString());
+
+          logWithMDC(logger, org.slf4j.event.Level.DEBUG, loggingDataMap, message);
+        } else {
+
+          logWithMDC(logger, org.slf4j.event.Level.INFO, loggingDataMap, message);
+        }
+      } catch (Exception e) {
+        logger.error("Error logging request: ", e);
+      }
+    }
+  }
+
+  static void logResponse(HttpResponse response, Logger logger, String message) {
+    if (logger.isInfoEnabled()) {
+      try {
+        Map<String, String> responseLogDataMap = new HashMap<>();
+        responseLogDataMap.put("response.status", String.valueOf(response.getStatusCode()));
+        responseLogDataMap.put("response.status.message", response.getStatusMessage());
+
+        Map<String, Object> headers = new HashMap<>();
+        response.getHeaders().forEach((key, val) -> headers.put(key, val));
+        responseLogDataMap.put("response.headers", headers.toString());
+        logWithMDC(logger, org.slf4j.event.Level.INFO, responseLogDataMap, message);
+      } catch (Exception e) {
+
+        logger.error("Error logging response: ", e);
+      }
+    }
+  }
+
+  static void logGenericData(GenericData genericData, Logger logger, String message) {
+    if (logger.isDebugEnabled()) {
+      try {
+
+        Map<String, String> contextMap = parseGenericData(genericData);
+        logWithMDC(logger, org.slf4j.event.Level.DEBUG, contextMap, message);
+      } catch (Exception e) {
+        logger.error("Error logging GenericData: ", e);
+      }
+    }
+  }
+
+  private static Map<String, String> parseGenericData(GenericData genericData) {
```

Comment:
I believe genericData can be null when we parse it from the response: https://github.com/googleapis/google-http-java-client/blob/b3f12421026ac3968e5682cf0cdd74a2eadae795/google-http-client/src/main/java/com/google/api/client/http/HttpResponse.java#L457-L459

We may need to guard against it.

Reply:
Thanks for the context. Yes, currently usage of parseGenericData() are both guarded with try-catch (e.g. L106 above).

Reply:
Using the L106 example above: If I'm understanding the flow above correctly, this would fail as a NPE when `genericData.forEach()` runs in L116.

I don't think an exception with NPE being returned is what we want to return to the user. I think we should catch the NPE and return a better error message: `No response body found in the response, etc.`

Additionally, I think this also would happen in the `parseResponseAs` method in ServiceAccountCredentials. An NPE would be caught and thrown back as an exception.

Reply:
This could be future improvements. For now, what I want to ensure is logging should not cause user application to crash.

Reply:
Ok, that's fine for now. Let's fix it in a future PR so we can provide more in-depth and helpful logs.